feat: Add a tutorial example for URL Shortner built using DiceDB and Go (#1259)
DiceDB/dice: docs/src/content/docs/tutorials/url-shortener.md

---
title: "Building a URL Shortener"
description: "Create a simple URL Shortener using DiceDB Go SDK."
---

This tutorial guides you through creating a URL shortener using DiceDB, a key-value store, with Go. We’ll set up endpoints to generate short URLs and redirect them to the original URLs.

Prerequisites

> 1. Go installed (at least version 1.18)
> 2. DiceDB server running locally

Reviewer: ```suggestion
1. Go (version 1.18 or later): [Download Go](https://golang.org/dl/)
2. DiceDB: A DiceDB server running locally. Refer to the [DiceDB Installation Guide](https://github.com/DiceDB/dice/pull/1259/get-started/installation) if you haven't set it up yet.
```

## Setup

> 1. Refer to [DiceDB Installation Guide](get-started/installation) to get your DiceDB server up and running with a simple Docker command.
> 2. Initialize a New Go Project
> 3. Install DiceDB Go SDK and other required packges.
>     ```bash
>     go get github.com/dicedb/dicedb-go
>     go get github.com/gin-gonic/gin
>     go get github.com/google/uuid
>     ```

Reviewer: ```suggestion
### 1. Install and Run DiceDB

Start a DiceDB server using Docker:

    ```bash
docker run -d -p 7379:7379 dicedb/dicedb
    ```

This command pulls the DiceDB Docker image and runs it, exposing it on port `7379`.

### 2. Initialize a New Go Project

Create a new directory for your project and initialize a Go module:

    ```bash
mkdir url-shortener
cd url-shortener
go mod init url-shortener
    ```

### 3. Install Required Packages

Install the DiceDB Go SDK and other dependencies:

    ```bash
go get github.com/dicedb/dicedb-go
go get github.com/gin-gonic/gin
go get github.com/google/uuid
    ```
```

> ## DiceDB Commands Used
> 
> Here are the main DiceDB commands we’ll use to store and retrieve URLs.
> 
> 1. `Set` Command: Stores a key-value pair in DiceDB.
> Syntax - `Set(key, value, expiration)`
> `key`: Unique identifier (e.g., short URL code)
> `value`: The data to store (serialized JSON)
> `expiration`: Optional; 0 means no expiration
> 
> 2. `Get` Command: Retrieves the value associated with a key.
> Syntax - `Get(key)`
> `key`: The identifier for the data to retrieve.

Reviewer: ```suggestion
## Understanding DiceDB Commands

We'll use the following DiceDB commands:

### `SET` Command

Stores a key-value pair in DiceDB.

- **Syntax**: `SET key value [expiration]`
  - `key`: Unique identifier (e.g., short URL code)
  - `value`: Data to store (e.g., serialized JSON)
  - `expiration`: Optional; time-to-live in seconds (use `0` for no expiration)

### `GET` Command

Retrieves the value associated with a key.

- **Syntax**: `GET key`
  - `key`: Identifier for the data to retrieve
```

> ## Code overview

Reviewer: ```suggestion
## Writing the Code

Create a file named `main.go` and add the following code:
```

    }

    // Redirects to the original URL based on the short URL ID
    func RedirectURL(c *gin.Context) {
        id := c.Param("id")

        // Retrieve stored URL data from DiceDB
        urlData, err := db.Get(context.Background(), id).Result()
        if err != nil {
            c.JSON(http.StatusNotFound, gin.H{"error": "URL not found"})
            return
        }

        // Deserialize JSON data back into URL struct
        var url URL
        if err := json.Unmarshal([]byte(urlData), &url); err != nil {
            c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode URL data"})
            return
        }

        // Redirect user to the original long URL
        c.Redirect(http.StatusFound, url.LongURL)
    }

    func main() {
        router := gin.Default()

        // Define endpoints for creating short URLs and redirecting
        router.POST("/shorten", CreateShortURL)
        router.GET("/:id", RedirectURL)

        // Start the server on port 8080
        if err := router.Run(":8080"); err != nil {
            log.Fatal("Failed to start server:", err)
        }
    }
    ```

> ## Explanation of Key Parts

Reviewer: ```suggestion
## Explanation
```

> 1. Database Initialization: The `init()` function sets up a DiceDB client to connect to `localhost:7379`.
> 2. `CreateShortURL` Endpoint: Handles the `/shorten` route. It generates a unique ID, constructs the short URL, serializes the URL data, and saves it in DiceDB.
> 3. `RedirectURL` Endpoint: Handles the `/:id` route. It retrieves the original URL by the short ID from DiceDB and redirects the user to it.
> 4. Starting the Server: The `main` function starts the Gin server on port `8080`.

Reviewer: We should make use of subheadings to make this clearer. PTAL at the following approach to write this:
```suggestion

### 1. Initialize the DiceDB Client

We set up the DiceDB client in the `init` function:

    ```go
db = dicedb.NewClient(&dicedb.Options{
    Addr: "localhost:7379",
})
    ```

### 2. Create Short URL Endpoint

- **Input Validation**: Ensures the `long_url` field is present.
- **Short ID Generation**: Uses `uuid` to create a unique 8-character ID.
- **Data Serialization**: Converts the `URL` struct to JSON.
- **Data Storage**: Saves the JSON data in DiceDB with the `Set` command.
- **Response**: Returns the generated short URL.

### 3. Redirect to Original URL Endpoint

- **Data Retrieval**: Fetches the URL data from DiceDB using the `Get` command.
- **Data Deserialization**: Converts JSON back to the `URL` struct.
- **Redirection**: Redirects the user to the `LongURL`.

### 4. Start the Server

The `main` function sets up the routes and starts the server on port `8080`.
```

> ## Starting the application server

Reviewer: ```suggestion
## Running the Application

### 1. Start the Go Application
```

1. Start the application
   ```bash
   go run main.go
   ```
   This will start the application server on port 8080 by default, you should see output similar to
   ```bash
   [GIN-debug] Listening and serving HTTP on :8080
   ```

> ## Interacting with the application
> 
> 1. Start DiceDB: Ensure DiceDB is running.

Reviewer: ```suggestion
### 2. Ensure DiceDB is Running

Ensure your DiceDB server is up and running on port `7379`.
```

> 2. Test the API:

Reviewer: ```suggestion
## Testing the Application

```

    - Shorten URL:

>         Send a POST request to `/shorten` with JSON body on Postman:
>         ```
>         {
>             "long_url": "https://example.com"
>         }
>         ```
> 
>         OR
> 
>         ```curl
>         curl -X POST -H "Content-Type: application/json" -d '{"long_url": "https://example.com"}' http://localhost:8080/shorten
>         ```

Reviewer: ```suggestion
**Using `curl`:**

    ```bash
curl -X POST -H "Content-Type: application/json" -d '{"long_url": "https://example.com"}' http://localhost:8080/shorten
    ```

**Response:**

    ```json
{
  "short_url": "http://localhost:8080/<short ID generated by the server>"
}
    ```
```

    - Redirect to Original URL:

>         Send a GET request to `/:id` with the short URL ID on Postman
> 
>         OR
> 
>         ```curl
>         curl -L http://localhost:8080/{short_id}
>         ```

Reviewer: ```suggestion
**Using `curl`:**

    ```bash
curl -L http://localhost:8080/abcd1234
    ```

**Using a Browser:**

Navigate to:

    ```
http://localhost:8080/abcd1234
    ```

You should be redirected to `https://example.com`.
```
